src/components/GoodsSearchProvider.php

```diff
@@ -0,0 +1,413 @@
+    {
+        $attributes = [];
+
+        foreach ($this->sortFields as $sort_field) {
+            $attributes[$sort_field] = [
+                'asc' => [$sort_field => SORT_ASC],
+                'desc' => [$sort_field => SORT_DESC],
+            ];
+        }
+
+        $sort = new Sort(
+            [
+                'attributes' => $attributes,
+                'sortParam' => $this->sortParameter
+            ]
+        );
+
+        return $sort;
+    }
+
+    private function getGoodsIds($query)
+    {
+        return $query->column();
+    }
+
+    private function getTree(&$full_section_array, &$sections)
+    {
+        foreach ($sections as $key => $section) {
+            if (array_key_exists($section['id'], $full_section_array)) {
+                $sections[$key]['subsections'] = $full_section_array[$section['id']];
+                $this->getTree($full_section_array, $sections[$key]['subsections']);
+                $full_section_array[$section['id']] = true;
+            }
+        }
+        return $sections;
+    }
+
+    private function getGoodsSectionsTree($goods_ids)
+    {
+        $category_goods_models = CategoryGoods::find()
+            ->select('structure_id')
+            ->addSelect('goods_id')
+            ->where(
+                [
+                    'goods_id' => $goods_ids
+                ]
+            )
+            ->all();
+
+        $category_goods = [];
+        foreach ($category_goods_models as $section) {
+            $category_goods[$section->structure_id][] = $section->goods_id;
+        }
+
+        $goods_category_models = BaseStructure::find()
+            ->innerJoin(
+                GoodsCategoryExtended::tableName(),
+                GoodsCategoryExtended::tableName() . '.[[model_id]] = ' . BaseStructure::tableName() . '.[[id]]'
+            )
+            ->where(
+                [
+                    'is_active' => 1,
+                    'is_deleted' => 0
+                ]
+            )
+            ->all();
+
+        $goods_category = [];
+
+        foreach ($goods_category_models as $section) {
+            $goods_category[$section->parent_id][] = [
+                'name' => $section->defaultTranslation->name,
+                'id' => $section->id,
+                'parentId' => $section->parent_id,
+                'slug' => $section->defaultTranslation->slug,
+                'count' => isset($category_goods[$section->id]) ? count($category_goods[$section->id]) : 0
+            ];
+        }
+
+        $first_category = reset($goods_category);
+
+        return $this->getTree($goods_category, $first_category);
+    }
+
+    private function getParentId()
+    {
+        $get = \Yii::$app->request->get();
+
+        return (
+        (!empty($get[$this->parentIdParameter]) && intval($get[$this->parentIdParameter]) > 0)
+            ? intval($get[$this->parentIdParameter])
+            : $this->parentId
+        );
+    }
+
+    private function getLimit()
```

Comment:
this method is excess.
use max and min limits in `pagination->pageSizeLimit` and pagination resolves all problems itself